test/infrastructure/container/docker.go

```diff
@@ -0,0 +1,358 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package container provides an interface for interacting with Docker and potentially
+// other container runtimes.
+package container
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"strings"
+
+	"github.com/docker/docker/api/types"
+	dockerContainer "github.com/docker/docker/api/types/container"
+	"github.com/docker/docker/api/types/network"
+	"github.com/docker/docker/client"
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/docker/go-connections/nat"
+)
+
+const (
+	httpProxy  = "HTTP_PROXY"
+	httpsProxy = "HTTPS_PROXY"
+	noProxy    = "NO_PROXY"
+)
+
+type docker struct {
+	dockerClient *client.Client
+}
+
+// NewDockerClient gets a client for interacting with a Docker container runtime.
+func NewDockerClient() (RuntimeInterface, error) {
+	dockerRuntime := &docker{}
+	dockerClient, err := getDockerClient()
+	if err != nil {
+		dockerRuntime.dockerClient = dockerClient
+	}
```

Comment:
Is there something you would recommend for handling differently? I could always assign anyway since it would be nil. Or switch is and return `nil, err` in case the consuming code somehow fails to check if `err` is nil in the response.

Reply:
I like it. 👍

Reply:
Hah, and even though I reread this several times, I now finally see that my logic was switched there. Doh!

Update coming shortly, after I make sure I'm not doing anything else blatantly wrong. 🤣